File: client/java/ircam/fts/client/FtsSocketConnection.java
//
// FTS client library
// Copyright (C) 2001 by IRCAM-Centre Georges Pompidou, Paris, France.
// 
// This program is free software; you can redistribute it and/or
// modify it under the terms of the GNU Lesser General Public License
// as published by the Free Software Foundation; either version 2
// of the License, or (at your option) any later version.
// 
// See file LICENSE for further informations on licensing terms.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU Lesser General Public License
// along with this program; if not, write to the Free Software
// Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
// 

package ircam.fts.client;

import java.io.*;
import java.net.*;

public class FtsSocketConnection extends FtsServerConnection {

  private static final int DEFAULT_PORT = 2023;
  private static final int DEFAULT_CONNECT_TIMEOUT = 30;

  /**
   *
   * Constructs a new FtsSocketConnection to the specified host and port.
   *
   * Tries to connect to FTS. Retries until connection is made or timed'out.
   *
   * @param hostname the host name
   * @param port the port number
   * @param connectTimeout the timeout value in seconds
   * @throws IOException if an I/O error occurs when creating the socket
   * @throws FtsClientException if a timeout occured when trying to connect
   * @throws UnknownHostException if the IP address of the host to which the connection is made could not be determined
   */
  public FtsSocketConnection( String hostname, int port, int connectTimeout) throws FtsClientException, IOException, UnknownHostException
  {
    do
      {
	try
	  {
	    socket = new Socket( hostname, port);
	  }
	catch( IOException e)
	  {
	  }

	if (socket != null)
	  break;

	try
	  {
	    Thread.sleep( 1000);
	  }
	catch (InterruptedException e)
	  {
	    throw new FtsClientException( "Connection interrupted");
	  }

	connectTimeout--;
      }
    while (connectTimeout > 0);

    if ( connectTimeout <= 0)
      throw new FtsClientException( "Cannot connect");

    in = socket.getInputStream();
    out = socket.getOutputStream();
  }

  public FtsSocketConnection( String hostname, int port) throws FtsClientException, IOException, UnknownHostException
  {
    this( hostname, port, DEFAULT_CONNECT_TIMEOUT);
  }

  public FtsSocketConnection( String hostname) throws FtsClientException, IOException, UnknownHostException
  {
    this( hostname, DEFAULT_PORT, DEFAULT_CONNECT_TIMEOUT);
  }  

  public FtsSocketConnection() throws FtsClientException, IOException, UnknownHostException
  {
    this( "127.0.0.1", DEFAULT_PORT, DEFAULT_CONNECT_TIMEOUT);
  }  

  public void close() throws IOException
  {
    socket.close();
    in = null;
    out = null;
  }

  public int read( byte[] b, int off, int len) throws IOException
  {
    return in.read( b, off, len);
  }

  public void write( byte[] b, int off, int len) throws IOException
  {
    out.write( b, off, len);
  }

  private Socket socket;
  private InputStream in;
  private OutputStream out;
}

File: client/cpp/src/FtsSocketConnection.cpp
//
// FTS client library
// Copyright (C) 2001 by IRCAM-Centre Georges Pompidou, Paris, France.
// 
// This program is free software; you can redistribute it and/or
// modify it under the terms of the GNU Lesser General Public License
// as published by the Free Software Foundation; either version 2
// of the License, or (at your option) any later version.
// 
// See file LICENSE for further informations on licensing terms.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU Lesser General Public License
// along with this program; if not, write to the Free Software
// Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
// 

#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <netdb.h>
#include <errno.h>
#include <unistd.h>
#include <fts/ftsclient.h>

namespace ircam {
namespace fts {
namespace client {

#ifndef WIN32
#define INVALID_SOCKET -1
#endif

  const int FtsSocketConnection::DEFAULT_PORT = 2023;
  const int FtsSocketConnection::DEFAULT_CONNECT_TIMEOUT = 30;

#if defined(WIN32)
  void FtsSocketConnection::initializeSocketLayer()
  {
    static int initializedSocketLayer = 0;

    if (initializedSocketLayer)
      return;

    initializedSocketLayer = 1;

    WORD wVersionRequested;
    WSADATA wsaData;
    int result;

    wVersionRequested = MAKEWORD(2, 2);
  
    result = WSAStartup( wVersionRequested, &wsaData );
    if (result != 0) {
      throw FtsClientException( "Couldn't initialize WinSock", result);
    }
  
    if (LOBYTE(wsaData.wVersion) != 2 || HIBYTE(wsaData.wVersion) != 2) {
      WSACleanup();
      throw FtsClientException( "Bad WinSock version");
    }
  }
#endif

  int FtsSocketConnection::connectOnce() throw( FtsClientException )
  {
    struct sockaddr_in server_addr;
    struct hostent *hostptr;
    struct in_addr addr;

    /* convert the host name to the inet address. call gethostbyname
       only when the host is not in a numbers-and-dots notation. this
       avoids a name resolution on windows machines */
    addr.s_addr = std::inet_addr( _hostname);

    if (addr.s_addr == INADDR_NONE) {

      /* host is not a numbers-and-dots notation. resolve the name. */
      hostptr = std::gethostbyname( _hostname);

      if ( !hostptr)
	throw FtsClientException( "Unknown host");

      addr = *(struct in_addr *)hostptr->h_addr_list[0];
    }

    _socket = std::socket( PF_INET, SOCK_STREAM, 0);

    if (_socket == INVALID_SOCKET)
      throw FtsClientException( "Can't create socket", errno);

    std::memset( &server_addr, 0, sizeof( server_addr));
    server_addr.sin_family = AF_INET;
    server_addr.sin_addr = addr;
    server_addr.sin_port = std::htons( _port);

    if (std::connect( _socket, (struct sockaddr *)&server_addr, sizeof(server_addr)) == 0)
      return 0;

#ifdef WIN32
    std::closesocket( _socket);
#else
    std::close( _socket);
#endif

    _socket = INVALID_SOCKET;

    return -1;
  }

  void FtsSocketConnection::connect() throw( FtsClientException )
  {
    int r, connectTimeout = _connectTimeout;

    do
      {
	if ((r = connectOnce()) == 0)
	  break;

	sleep( 1);
	connectTimeout--;
      }
    while (connectTimeout > 0);

    if ( r < 0)
      throw FtsClientException( "Cannot connect", errno);
  }

  FtsSocketConnection::FtsSocketConnection( const char *hostname, int port, int connectTimeout) throw( FtsClientException)
    : _socket( INVALID_SOCKET), _hostname( hostname), _port( port), _connectTimeout( connectTimeout)
  {
#ifdef WIN32
    initializeSocketLayer();
#endif

    connect();
    startThread();
  }

  void FtsSocketConnection::close() throw (FtsClientException)
  {
  }

  int FtsSocketConnection::read( unsigned char *b, int len) throw (FtsClientException)
  {
    int r;

    r = std::read( _socket, b, len);
    
    if (r == -1)
      throw FtsClientException( "Error in message receiving", errno);
    else if (r == 0)
      throw FtsClientException( "Socket closed");

    return r;
  }

  void FtsSocketConnection::write( const unsigned char *b, int len) throw (FtsClientException)
  {
    if ( std::write( _socket, b, len) < 0)
      {
	throw FtsClientException( "Error in sending message", errno);
      }
  }

};
};
};
